tests/test_ga4_conversion_window_invalid.py

```diff
@@ -0,0 +1,147 @@
+"""Test tap configurable properties. Specifically the conversion_window"""
+import os
+import uuid
+from datetime import datetime as dt
+from datetime import timedelta
+
+from tap_tester import menagerie, connections, runner, LOGGER
+
+from base import GA4Base
+
+
+class ConversionWindowInvalidTest(GA4Base):
+    """
+    Test tap's sync mode throws expected exception with invalid conversion_window values set.
+    Validate setting the conversion_window configurable property.
+
+    Test Cases:
+
+    Verify tap throws critical error when a value is provided directly by a user which is
+    outside the set of acceptable values.
+
+      Acceptable values: { 30, 60, 90 } type = string
+    """
+    conversion_window = ''
+    conversion_window_type = ''
+
+
+    @classmethod
+    def name(cls):
+        return f"tt_ga4_conv_window_invalid_{cls.conversion_window_type}"
+
+    def get_properties(self, original: bool = True):
+        # Use the same UUID for each custom report
+        if not self.custom_report_id_1 and not self.custom_report_id_2:
+            type(self).custom_report_id_1 = str(uuid.uuid4())
+            type(self).custom_report_id_2 = str(uuid.uuid4())
+
+        return_value = {
+            'start_date': (dt.utcnow() - timedelta(days=3)).strftime(self.START_DATE_FORMAT),
+            'conversion_window': self.conversion_window,
+            'property_id': os.getenv('TAP_GA4_PROPERTY_ID'),
+            'account_id': '659787',
+            'oauth_client_id': os.getenv('TAP_GA4_CLIENT_ID'),
+            'user_id': os.getenv('TAP_GA4_USER_ID'),
+            'report_definitions': [
+                {"id": self.custom_report_id_1, "name": "Test Report 1"},
+                {"id": self.custom_report_id_2, "name": "Test Report 2"},
+            ]
+        }
+        if original:
+            return return_value
+
+        if self.start_date:
+            return_value["start_date"] = self.start_date
+        if self.request_window_size:
+            return_value["request_window_size"] = self.request_window_size
+        return return_value
+
+
+    def streams_to_test(self):
+        # testing all streams creates massive quota issues
+        custom_id = self.custom_reports_names_to_ids()['Test Report 1']
+        return {
+            'content_group_report',
+            custom_id
+        }
+
+
+    def run_test(self):
+        """
+        Testing that basic sync functions without Critical Errors when
+        a valid conversion_window is set.
+        """
+
+        try:
+            # Create a connection
+            conn_id = connections.ensure_connection(self, original_properties=False)
```

Comment:
I'm actually curious why we would be able to create a connection that is incorrectly configured.  Shouldn't this be rejected at this point as not a valid value?

Actually now that I see this is wrapped in a try maybe it does fail here.  Instead of a try catch we should do something like this for the thing that throws the exception and test the message this way.

with self.assertRaises(Exception) as context:
            
         broken_function()
  self.assertTrue('This is broken' in context.exception)

There is so much in this try I don't really know where it supposed to be thrown.

Reply:
That is the expectation in the test.  It does not call out exactly where it expects to fail but I believe the ensure connection is the place.  Will refactor with this approach.  May need some discussion :)